packages/lib/src/basic-io.ts

```diff
@@ -12,23 +12,35 @@ import { TableFieldDefinition } from './table-generator'
 export type GetDataFunction<O> = () => Promise<O>
 export type ListDataFunction<L> = () => Promise<L[]>
 export type LookupDataFunction<ID, O> = (id: ID) => Promise<O>
-export type ExecuteCommandFunction<I, O> = (input: I) => Promise<O>
+export type ActionFunction<I, O> = (input: I) => Promise<O>
+export type IdTranslationFunction<ID, L> = (idOrIndex: ID | string, listFunction: ListDataFunction<L>) => Promise<ID>
+export type IdRetrievalFunction<ID, L> = (primaryKeyName: string, list: L[]) => Promise<ID>
 
 
+export interface Sorting {
+	primaryKeyName: string
+	sortKeyName: string
+}
+export interface Naming {
+	itemName?: string
+	pluralItemName?: string
+}
```

Comment:
The "nothing found" messages before were just "no items" found which is all fine and dandy if you're just querying, say `smartthings devices` or `smartthings installedapps` and then get "no items found" but I was worried it would be too confusing when you `smartthings installedapps:rename` (without specifying an item to rename) and then get `no items found`.

Reply:
Since it matches the behavior of other commands, I think it is fine to say no items found when renaming. The user can always reference the `--help` output. If it helps remove some complexity from functions having to deal with this special type.

Reply:
Are you saying you think I should remove this code that allows us a more specific message or did you misunderstand the reason behind my comment?

Reply:
I think we should remove the extra naming code. I don't see a strong case for it when we already have several cases using `no items found`, so it feels consistent to me.

Reply:
I wrote it in such a way that we can start including `itemName` (and optionally `pluralItemName` so we get more specific messages in other places very easily as well. Here's what this looks like in practice:

```
ST ~/dev/cli/smartthings-cli $ smartthings installedapps:rename
no installed apps found
```
vs.
```
ST ~/dev/cli/smartthings-cli $ smartthings installedapps:rename
no items found
```
To me just using "items" might be more consistent with other places we do that but it's inconsistent with making the CLI easier to use. Contrast this to what we do when there are items found...we display them and let the user pick one.

I can remove it if you really have strong feelings but I do think it's helpful.